File: DataStructures/Heaps/qHeapSolution.js
function processData(input) {
    //Enter your code here
    const data = input.split('\n');
    let myHeap = [];
    const numberOfqueries = data.shift();
    for (let i = 0; i < numberOfqueries; i++) {
        const option = parseInt(data[i])
        let item = parseInt(data[i].substr(2));

        if (option == 1) {
            myHeap.push(item);
        } else if (option == 2) {
            index = myHeap.indexOf(item)
            myHeap.splice(index, 1);
        } else {
            let minimum = Infinity;
            for (let j = 0; j < myHeap.length; j++){
                if (myHeap[j] < minimum) {
                    minimum = myHeap[j];
                }
            }
            console.log(minimum);
        }
    }
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
    processData(_input);
});

File: DataStructures/Heaps/qheap_solution.py
import heapq
data = int(input())
heap = []
dictionary = {}
for i in range(data):
    option = input().split()
    if option[0] == '1':
        option = int(option[1])
        dictionary[option] = 0
        heapq.heappush(heap, option)
    elif option[0] == '2':
        option = int(option[1])
        dictionary[option] = 1
    else:
        while dictionary[heap[0]] == 1:
            heapq.heappop(heap)
        print(heap[0])